src/Microsoft.AspNetCore.Http.Connections.Client/Internal/WebSocketsTransport.cs

```diff
@@ -181,8 +181,23 @@ public WebSocketsTransport(HttpOptions httpOptions, ILoggerFactory loggerFactory
             {
                 while (true)
                 {
-                    var memory = _application.Output.GetMemory();
+#if NETCOREAPP2_1
+                    // Do a 0 byte read so that idle connections don't allocate a buffer when waiting for a read
+                    var result = await socket.ReceiveAsync(Memory<byte>.Empty, CancellationToken.None);
+
+                    if (result.MessageType == WebSocketMessageType.Close)
+                    {
+                        Log.WebSocketClosed(_logger, _webSocket.CloseStatus);
+
+                        if (_webSocket.CloseStatus != WebSocketCloseStatus.NormalClosure)
+                        {
+                            throw new InvalidOperationException($"Websocket closed with error: {_webSocket.CloseStatus}.");
+                        }
 
+                        return;
+                    }
+#endif
+                    var memory = _application.Output.GetMemory();
 #if NETCOREAPP2_1
                     var receiveResult = await socket.ReceiveAsync(memory, CancellationToken.None);
```

Comment:
Put a comment here that we're guaranteed not to get a close status because the zero-byte read didn't return a close frame.